### PR TITLE
CP-25081: UI improvements for USB passthrough

### PR DESCRIPTION
**Pull request by michael2012z**
This PR covers 4 items in CP-25081:
- USB Tab: The position and size of the devices grid on the USB tab does not match the other pages (e.g. Storage, NICs)
- VM Properties: We should have some text on this page above the devices grid
- Should be able to disable USB passthrough if the device is attached to a VM?
- In USB list for both Host and VM, when PUSB description is missing (this happens on special devices), show "Vendor:  <vid>; Product: <pid>" instead (vid and pid can never be empty).

### CODE REVIEW

**Comment by huizh** [comment]
Hi @MihaelaStoica, please could you review this PR?

Thanks.

File: XenModel/Messages.resx [hunk 1]
```diff
@@ -3934,6 +3934,9 @@ This will also delete its subfolders.</value>
   <data name="DEVICE_POSITION_SCANNING" xml:space="preserve">
     <value>Scanning for free device positions</value>
   </data>
+  <data name="DIALOG_ATTACH_USB_NO_DEVICES_AVAILABLE" xml:space="preserve">
+    <value>No Devices Available</value>
```
Comment (MihaelaStoica) [inline XenModel/Messages.resx]: "No devices available"

File: XenAdmin/SettingsPanels/USBEditPage.cs [hunk 2]
```diff
@@ -273,7 +273,9 @@ public void UpdateDetails()
                 PUSB pusb = _vusb.Connection.Resolve(usbgroup.PUSBs[0]);
 
                 locationCell.Value = pusb.path;
-                descriptionCell.Value = pusb.description;
+                descriptionCell.Value = string.IsNullOrEmpty(pusb.description) ? 
+                        string.Format("Vendor: {0}; Product: {1}", pusb.vendor_id, pusb.product_id) :
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/SettingsPanels/USBEditPage.cs]: pusb.description is made up of vendor_desc and product_desc. A missing pusb.description indicates empty vendor_desc and product_desc.

File: XenAdmin/TabPages/UsbPage.cs [hunk 3]
```diff
@@ -172,18 +172,28 @@ internal void UnregisterHandlers()
...
 
             if (selectedRow != null && selectedRow.Pusb.passthrough_enabled)
+            {
+                PUSB pusb = selectedRow.Pusb;
+                USB_group usbGroup = pusb.Connection.Resolve(pusb.USB_group);
+                bool attached = (usbGroup != null) && (usbGroup.VUSBs != null) && (usbGroup.VUSBs.Count > 0);
+
                 buttonPassthrough.Text = Messages.USBLIST_DISABLE_PASSTHROUGH_HOTKEY;
+
+                if (attached)
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/TabPages/UsbPage.cs]: Thanks.

File: XenModel/Messages.resx (hunk 1 above)
Comment (michael2012z) [reply XenModel/Messages.resx]: OK.

File: XenAdmin/TabPages/UsbPage.cs [hunk 4]
```diff
@@ -238,7 +248,9 @@ public PUSB Pusb
             public void UpdateDetails()
             {
                 locationCell.Value = _pusb.path;
-                descriptionCell.Value = _pusb.description;
+                descriptionCell.Value = string.IsNullOrEmpty(_pusb.description) ?
+                        string.Format("Vendor: {0}; Product: {1}", _pusb.vendor_id, _pusb.product_id) :
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/TabPages/UsbPage.cs]: Thanks for sharing. Partial-class is interesting and useful in class extension.

File: XenAdmin/Dialogs/PropertiesDialog.cs [hunk 5]
```diff
@@ -226,7 +226,14 @@ private void Build()
 
                 if (is_vm && !Helpers.FeatureForbidden(xenObjectCopy, Host.RestrictUsbPassthrough))
                 {
-                    ShowTab(usbEditPage = new USBEditPage { VerticalTabs = verticalTabs });
+                    foreach (Host host in pool.Connection.Cache.Hosts)
```
[the remark this answers is not in the document]
Comment (michael2012z) [reply XenAdmin/Dialogs/PropertiesDialog.cs]: Need to build a List from the array at first, like:
if (new List<Host>(pool.Connection.Cache.Hosts)).Exists(host => host.PUSBs.Count > 0)

[the remark this answers is not in the document]
Comment (MihaelaStoica) [reply XenAdmin/Dialogs/PropertiesDialog.cs]: You can use `Any` as above, just need to add a `using` directive for the `System.Linq` namespace

**Pull request merged**